blockchain/storagev2/storage_write.go

```diff
@@ -0,0 +1,105 @@
+package storagev2
+
+import (
+	"math/big"
+
+	"github.com/0xPolygon/polygon-edge/helper/common"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+func (w *Writer) PutHeader(h *types.Header) {
+	// block_num_u64 + hash -> header (RLP)
+	w.putRlp(HEADER, getKey(h.Number, h.Hash), h)
+}
+
+func (w *Writer) PutBody(bn uint64, bh types.Hash, body *types.Body) {
+	// block_num_u64 + hash -> body (RLP)
+	w.putRlp(BODY, getKey(bn, bh), body)
+}
+
+func (w *Writer) PutHeadHash(h types.Hash) {
+	w.putIntoTable(HEAD_HASH, HEAD_HASH_KEY, h.Bytes())
+}
+
+func (w *Writer) PutHeadNumber(bn uint64) {
+	w.putIntoTable(HEAD_NUMBER, HEAD_NUMBER_KEY, common.EncodeUint64ToBytes(bn))
+}
+
+func (w *Writer) PutTxLookup(hash types.Hash, bn uint64) {
+	w.putIntoTable(TX_LOOKUP, hash.Bytes(), common.EncodeUint64ToBytes(bn))
+}
+
+func (w *Writer) PutBlockLookup(hash types.Hash, bn uint64) {
+	w.putIntoTable(BLOCK_LOOKUP, hash.Bytes(), common.EncodeUint64ToBytes(bn))
+}
+
+func (w *Writer) PutReceipts(bn uint64, bh types.Hash, receipts []*types.Receipt) {
+	rs := types.Receipts(receipts)
+	w.putRlp(RECEIPTS, getKey(bn, bh), &rs)
+}
+
+func (w *Writer) PutCanonicalHeader(h *types.Header, diff *big.Int) {
+	w.PutHeader(h)
+	w.PutHeadHash(h.Hash)
+	w.PutHeadNumber(h.Number)
+	w.PutBlockLookup(h.Hash, h.Number)
+	w.PutCanonicalHash(h.Number, h.Hash)
+	w.PutTotalDifficulty(h.Number, h.Hash, diff)
+}
+
+func (w *Writer) PutCanonicalHash(bn uint64, hash types.Hash) {
+	w.putIntoTable(CANONICAL, common.EncodeUint64ToBytes(bn), hash.Bytes())
+}
+
+func (w *Writer) PutTotalDifficulty(bn uint64, bh types.Hash, diff *big.Int) {
+	w.putIntoTable(DIFFICULTY, getKey(bn, bh), diff.Bytes())
+}
+
+func (w *Writer) PutForks(forks []types.Hash) {
+	fs := Forks(forks)
+	w.putRlp(FORK, FORK_KEY, &fs)
+}
+
+func (w *Writer) putRlp(t uint8, k []byte, raw types.RLPMarshaler) {
+	var data []byte
+
+	if obj, ok := raw.(types.RLPStoreMarshaler); ok {
+		data = obj.MarshalStoreRLPTo(nil)
+	} else {
+		data = raw.MarshalRLPTo(nil)
+	}
+
+	w.putIntoTable(t, k, data)
+}
+
+func (w *Writer) putIntoTable(t uint8, k []byte, data []byte) {
+	w.getBatch(t).Put(t, k, data)
+}
+
+func (w *Writer) WriteBatch() error {
+	for i, b := range w.batch {
+		if b != nil {
+			err := b.Write()
```

Comment:
Is there a way we can make this writing to two different databases atomic? Each batch write is atomic, but now we have two different writes (`for i, b := range w.batch`)

Reply:
It seems to me that all the data needed for the index database are present in the main database. Therefore, we can likely regenerate the index data for blocks where writing the batch index failed. Probably on node (re)start. Am I correct?

Reply:
We can't make it atomic if we are looking at the database level, but method doing write is atomic, so it will return an error if any of those writes fails.
Regarding indexes, we don't have tx lookup data in the main database. The same for block (if we are getting block with hash). However IMO, after I measured performances with one database and having in mind how leveldb doing write, I think we can go with one database but keep this way of doing write.

Reply:
Just for the record, although 1 database solution with this way of writing will have the same speed performances as 2 databases, number of read and written bytes from/to the disc will be much higher because we will have 1 merge for each main DB key/value pair and with 2 databases this merge is completely avoided. In order to complete that merge leveldb must first read data from the disc and then write it again. Merge is finding appropriate position based on key value in L1+ files. With 2 databases L0 files can be moved into L1 without any merging with existing L1 files because L0 file contains sorted key/value pairs which goes at the end of L1.

Reply:
"Regarding indexes, we don't have tx lookup data in the main database. The same for block (if we are getting block with hash). Hence we need 2 databases, but lookup database can be regenerated completely from the main DB. "
That's important for me. We can regenerate if needed. I know this should not happened in 99,99% cases but we need to have recovering mechanism in those rare cases

Reply:
Yes, we can regenerate if needed.